KAFKA-14104: Add CRC validation when iterating over Metadata Log Records

niket-goel opened the pull request:
This commit adds a check to ensure the RecordBatch CRC is valid when
iterating over a Batch of Records using the RecordsIterator. The
RecordsIterator is used by both Snapshot reads and Log Records reads in
Kraft. The check can be turned off by a class parameter and is on by default.

jsancio commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1216,7 +1216,16 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       case FetchTxnCommitted => fetchLastStableOffsetMetadata
     }
-    localLog.read(startOffset, maxLength, minOneMessage, maxOffsetMetadata, isolation == FetchTxnCommitted)
+    val fetchDataInfo = localLog.read(startOffset, maxLength, minOneMessage, maxOffsetMetadata, isolation == FetchTxnCommitted)
```
Does this mean that Kafka needs to read all of the batches to handle a Fetch request? For performance, Kafka doesn't read the batches when handling a Fetch request. Kafka sends bytes. It is up to the reader (Consumer or Follower) to validate the bytes received.

niket-goel replied:
That is a fair question. I was contemplating that too. The alternative of having a reader verify the bytes places less stress on the log layer in Kafka, but is not really fool proof.
e.g. in that scheme the listeners that are reading the batches (both for controller and broker) would do the CRC check (assuming the batch header reaches that layer).
Actually now that I think about it, does having the check in the `RecordsIterator` protect all readers?

jsancio commented on raft/src/main/java/org/apache/kafka/snapshot/RecordsSnapshotReader.java:
```diff
@@ -106,6 +106,7 @@ public static <T> RecordsSnapshotReader<T> of(
         BufferSupplier bufferSupplier,
         int maxBatchSize
     ) {
+        // TODO: Is this a good place to perform delimeter check (i.e. existence of header and footer?)
```
If you do this here, that means you have to read the snapshot twice, right?

niket-goel replied:
That is the downside, yes. We could also seek to the end of the snapshot file and just read the first and last batch. More ideal w.r.t performance would be to just check for the existence of the footer as you are iterating through the snapshot. You would still realize that you have an incomplete snapshot, but you would have applied a part of that incomplete snapshot to your memory (not sure if that poses a correctness risk).

jsancio reviewed (commented):
RecordsIterator is used to read cluster metadata log and snapshot.

niket-goel committed:
KAFKA-14104: Add CRC validation when iterating over Metadata Log Records

This commit adds a check to ensure the RecordBatch CRC is valid when
iterating over a Batch of Records using the RecordsIterator. The
RecordsIterator is used by both Snapshot reads and Log Records reads in
Kraft. The check can be turned off by a class parameter and is on by default.

niket-goel commented:
The build failures are a little cryptic. The same tests all pass locally, and the failing step in the build only complains of the script exiting with a non-zero status, but there are no actual failing tests. Will just try to trigger another run.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/RecordsIterator.java:
```diff
@@ -179,8 +182,13 @@ private Optional<Batch<T>> nextBatch() {
-    private Batch<T> readBatch(DefaultRecordBatch batch) {
+    private Batch<T> readBatch(DefaultRecordBatch batch) throws CorruptRecordException {
         final Batch<T> result;
+        if (doCrcValidation) {
```
You can move this block before `final Batch<T> tresult` and keep `result` close to where it is initialized.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/RecordsIterator.java:
```diff
@@ -179,8 +182,13 @@ private Optional<Batch<T>> nextBatch() {
+    private Batch<T> readBatch(DefaultRecordBatch batch) throws CorruptRecordException {
         final Batch<T> result;
+        if (doCrcValidation) {
+            // Perform a CRC validity check on this block.
```
typo; did you mean "on this batch."?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/RecordsIterator.java:
```diff
@@ -49,17 +50,20 @@
     // Number of bytes from records read up to now
     private int bytesRead = 0;
     private boolean isClosed = false;
+    private boolean doCrcValidation = false;
```
This should be `private final boolean doCrcValidation;`. It should be moved with the rest of the `final` fields.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/RecordsIteratorTest.java:
```diff
@@ -85,18 +88,41 @@ public void testFileRecords(
+        List<TestBatch<String>> batches = createBatches(seed);
+        MemoryRecords memRecords = buildRecords(compressionType, batches);
+        // Corrupt the record buffer
+        memRecords.buffer().putInt(DefaultRecordBatch.LAST_OFFSET_DELTA_OFFSET, new Random(seed).nextInt());
```
Does a random int always corrupt the batch? Meaning, it is possible that this test sometime fails because it was unlucky to pick a random int that didn't invalidate the CRC.

Can you explain in a comment why this buffer/memory change specifically?

niket-goel replied:
There is a non-zero probability that the test might fail due to the random int colliding with the actual value. Will modify the test to fix this.
As for the field, it was one of the two exposed offsets that I could choose to corrupt. I am actually not happy about this either. I am going to expose the CRC_OFFSET as a public field and corrupt that instead. Should make for a more reliable test.

jsancio reviewed (commented):
Thanks for the changes @niket-goel

niket-goel committed:
Addressing review comments

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/RecordsIteratorTest.java:
```diff
@@ -98,8 +98,12 @@ public void testCrcValidation(
+        // Read the Batch CRC for the first batch from the buffer
+        ByteBuffer readBuf = memRecords.buffer().asReadOnlyBuffer();
+        readBuf.position(DefaultRecordBatch.CRC_OFFSET);
+        Integer actualCrc = readBuf.getInt();
```
How about this to avoid the boxing and object creation:
```java
int actualCrc = readBuf.getInt();
```

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/RecordsIteratorTest.java:
```diff
@@ -110,6 +114,15 @@ public void testCrcValidation(
+
+        // Verify check does not trigger when the corruption is fixed
+        assertDoesNotThrow(() -> testIterator(batches, memRecords, true));
+        FileRecords moreFileRecords = FileRecords.open(TestUtils.tempFile());
```
This also applies to existing tests. The tests should close the `FileRecords` objects when they are done using them.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/RecordsIteratorTest.java:
```diff
@@ -98,8 +98,12 @@ public void testCrcValidation(
-        // Corrupt the record buffer
-        memRecords.buffer().putInt(DefaultRecordBatch.LAST_OFFSET_DELTA_OFFSET, new Random(seed).nextInt());
+        // Read the Batch CRC for the first batch from the buffer
+        ByteBuffer readBuf = memRecords.buffer().asReadOnlyBuffer();
```
Btw, `adReadOnlyBuffer` is probably not needed since `MemoryBuffer#buffer` returns a duplicate view into the buffer.

niket-goel committed:
Some test improvements based on feedback

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/RecordsIterator.java:
```diff
@@ -179,7 +184,12 @@ private Optional<Batch<T>> nextBatch() {
     }
 
-    private Batch<T> readBatch(DefaultRecordBatch batch) {
+    private Batch<T> readBatch(DefaultRecordBatch batch) throws CorruptRecordException {
```
`CorruptRecordException` is a runtime exception and doesn't need to be added to the set of checked exceptions.

jsancio reviewed (commented):
Should be good to merge after this comment.

niket-goel committed:
Minor: removing a check for a runtime exception

jsancio reviewed (approved):
LGTM

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/RecordsIterator.java:
```diff
@@ -180,6 +184,11 @@ private Optional<Batch<T>> nextBatch() {
     private Batch<T> readBatch(DefaultRecordBatch batch) {
+        if (doCrcValidation) {
+            // Perform a CRC validity check on this batch
+            batch.ensureValid();
```
This throws an exception that doesn't include a stacktrace. This is a rare event so we want to pay the cost of generating a stacktrace as it will help us debug the issue.
